spring-xd-dirt/src/main/java/org/springframework/xd/dirt/stream/StreamDeployer.java

```diff
@@ -86,4 +105,59 @@ protected String getDeploymentPath(StreamDefinition definition) {
 		return Paths.build(Paths.STREAM_DEPLOYMENTS, definition.getName());
 	}
 
+	/**
+	 * The migration code to run against existing stream definitions that don't have module definitions set already.
+	 * The following method will update the module definitions for those stream definitions.
+	 * See https://jira.spring.io/browse/XD-2854
+	 */
+	private void updateModuleDefinitions() {
+		if (this.parser != null && this.zkConnection != null && this.zkConnection.getClient() != null) {
+			try {
+				CuratorFramework client =  this.zkConnection.getClient();
+				if (client.checkExists().forPath(Paths.STREAMS) != null) {
+					ObjectWriter objectWriter =
+							new ObjectMapper().writerWithType(new TypeReference<List<ModuleDefinition>>() {});
+					Iterable<StreamDefinition> streamDefinitions = findAll();
+					for (StreamDefinition definition : streamDefinitions) {
+						String streamName = definition.getName();
+						String path = Paths.build(Paths.STREAMS, streamName);
+						try {
+							byte[] bytes = client.getData().forPath(path);
+							if (bytes != null) {
+								Map<String, String> map = ZooKeeperUtils.bytesToMap(bytes);
+								if (map.get(MODULE_DEFINITIONS_KEY) == null) {
+									List<ModuleDescriptor> moduleDescriptors = this.parser.parse(streamName,
+											definition.getDefinition(), definitionKind);
+									List<ModuleDefinition> moduleDefinitions = createModuleDefinitions(moduleDescriptors);
+									if (!moduleDefinitions.isEmpty()) {
+										map.put(DEFINITION_KEY, definition.getDefinition());
+										try {
+											map.put(MODULE_DEFINITIONS_KEY,
+													objectWriter.writeValueAsString(moduleDefinitions));
+										}
+										catch (JsonProcessingException jpe) {
```

Comment:
Yeah, we can skip the stream as the `streamDefinition` key should have been there in ZK data.